### PR TITLE
[2017.2] Controller Input and Rendering/glTF Updates

### PR DESCRIPTION
**Pull request by keveleigh**
Refactored the ControllerVisualizer and ControllerDebug scripts. Created a "smarter" way to load the glTF stream, making it more resistant to changes on the remote loader. Updated the glTF files from the remote repo (perhaps a submodule, once the remote will build for UWP without any edits, would be better).

### CODE REVIEW

File: Assets/HoloToolkit/Input/Scripts/ControllerInfo.cs [hunk 1]
```diff
@@ -219,5 +217,119 @@ public void LoadInfo(Transform[] childTransforms, ControllerVisualizer visualize
...
             }
         }
+
+        public void AnimateGrasp(bool isGrasped)
+        {
+            if (grasp != null && graspPressed != null && graspUnpressed != null && isGrasped != wasGrasped)
+            {
+                if (isGrasped)
+                {
+                    grasp.transform.localPosition = graspPressed.localPosition;
```
Comment (NeerajW) [inline Assets/HoloToolkit/Input/Scripts/ControllerInfo.cs]: I see this pattern a lot in the code. You could consider doing something like:

private void SetButtonPositonRotation(GameObject controllerButton, Transform controllerButtonTransform)
{
controllerButton.transform.localPosition = controllerButtonTransform.localPosition;
rotation blah
}

You could make this even better by passing in structs/class for each button type as most are similar calls.

Comment (keveleigh) [inline Assets/HoloToolkit/Input/Scripts/ControllerInfo.cs]: Ooh good idea! It was definitely feeling a bit repetitive, but I hadn't thought of that.

**Comment by jessemcculloch** [comment]
@NeerajW - You might have to push this one through without a second reviewer.  Not sure any of us have a set of motion controllers to test with.

**Comment by StephenHodgson** [comment]
@NeerajW I just got some controllers today. I'll test this soon.

**Comment by jessemcculloch** [comment]
@StephenHodgson - Lucky! I can't wait to try them.  You'll have to let me know how they are!

File: ProjectSettings/ProjectVersion.txt [hunk 2]
```diff
@@ -1 +1 @@
-m_EditorVersion: 2017.2.0b4
+m_EditorVersion: 2017.2.0b6
```
Comment (StephenHodgson) [inline ProjectSettings/ProjectVersion.txt]: Did you mean to update this with this PR?

**Comment by StephenHodgson** [comment]
I guess my real question is why aren't we just importing these files into the editor and calling it a day instead of messing around with glTF at runtime?

**Comment by jessemcculloch** [comment]
I can't speak officially to this, but I was under the impression that the Crystal Key Controllers are reference designs, and in theory we could/should see other designs come from the OEM's as well.  Be using the gITF at runtime, it would allow the correct controller model to render per OEM controller without us having to find and add each model to the HTK.

**Comment by NeerajW** [comment]
@jessemcculloch said it perfectly. OEMs will also ship drivers via WU which can be read in via glTF which is soon becoming the VR standard for model rendering.

It would be less than ideal if we had to bake in models for all different types of controllers that would work on mixed reality.

**Comment by StephenHodgson** [comment]
Ah I see.

Then is there any way we can use the build in JSON utitlity in Unity instead of the Newtonsoft dependencies?

**Comment by keveleigh** [comment]
@StephenHodgson That dependency comes from https://github.com/AltspaceVR/UnityGLTF.

I looked *very* briefly into converting it, but there seemed to be a lot of work there in order to remove that dependency.

**Comment by StephenHodgson** [comment]
Should be pretty straight forward if the JSON keys are known ahead of time.  All you have to do is write the struct or class the JSON values can deserialize to.

**Comment by StephenHodgson** [comment]
I'm just worried about adding more dependencies to the repository.

**Comment by NeerajW** [comment]
The dependencies are in the API so I am not concerned there if we take glTF as well. 
WU = Windows Update.

Jason serialization is a neat idea but not a blocker here. This glTF code can be used for generic model rendering as well.

**Comment by NeerajW** [comment]
@keveleigh Should we consider sub module-ing glTF?

**Comment by StephenHodgson** [comment]
I'm also not a fan of adding a submodule.  This adds a level of complexity to the repo that could make working with it even more difficult if done improperly.

**Comment by keveleigh** [comment]
I think once the [source repo](https://github.com/KhronosGroup/UnityGLTF) "just works" as far as building for UWP, we should revisit the submodule idea. It does add an extra step for people pulling the repo for the first time though, and I'm not sure how well it plays with downloading via GitHub's generated zip.

**Comment by NeerajW** [comment]
Agreed. For now no submodule.

**Comment by StephenHodgson** [comment]
How to test?

**Comment by keveleigh** [comment]
@StephenHodgson The MotionControllerTest scene is the best starting point. That should visualize the left controller as a rectangle when launched either through the Editor or via deploy.

**Comment by StephenHodgson** [comment]
So it doesn't render the controller models?

**Comment by keveleigh** [comment]
@StephenHodgson Ahh yes, I was very recently informed that the model driver hasn't actually been distributed outside Microsoft yet...I was under the impression that it had been :/

**Comment by StephenHodgson** [comment]
> Ahh yes, I was very recently informed that the model driver hasn't actually been distributed outside Microsoft yet..

Yeah I wasn't able to get this to work. I'll try updating the drivers on the controllers to see if maybe that'll help.

**Comment by jessemcculloch** [comment]
So, to add another opinion to the submodule thing, I am actually against it.  Right now there is no barrier to entry for someone to pull the zip file.  Someone with no idea how to use Git can still move forward.  Once you start using submodules, you take that away, and force people to clone the repo and then pull the submodules.  I don't think we gain enough by using submodules to justify it.  

FYI, I just spend 30 minutes messing around with the MRDL repo trying to figure out why stuff was missing before realizing they had implemented submodules....

**Comment by StephenHodgson** [comment]
Any update as to when the drivers for this will roll out?

**Comment by keveleigh** [comment]
@StephenHodgson From the latest info I've been given, it should be publicly released now. Please check Device Manager -> Human Interface Devices -> Motion Controller -> Driver for at least version 16249.

image

**Comment by StephenHodgson** [comment]
Yeah I've got 16269 so I should be good

**Comment by StephenHodgson** [comment]
When running in the editor I wasn't able to get the override models to show properly.

**Comment by StephenHodgson** [comment]
I was unable to build and run, because the `UnityPlayer.dll` keeps throwing an Access Violation Exception.

**Comment by keveleigh** [comment]
@StephenHodgson 

Yeah, that'll need to change in the main UnityGLTF repo as well. I guess we can change them in our branch for now first.

Sometimes the Editor can be finicky getting SourceDetected events. I'm not sure a great way to force it to happen.

I've never seen this issue, weird.

**Comment by StephenHodgson** [comment]
Yeah, I tried turning the controllers on and off to force the `SourceDetected` events, but no dice.

**Comment by keveleigh** [comment]
@StephenHodgson 

> I know we're taking from the other repository but it's obvious they're not using Unity.

The repo is a Unity-specific glTF loader. It still in active development, so I guess they haven't gotten around to fixing their warnings yet.

Yep, if the override models are specified, the actual models will never load. You'll need to remove the overrides. The rotation of the controller models sounds like a known bug: the rotation will always be from the reference of the world's forward direction, based upon when you centered your space during mixed reality setup.

**Comment by StephenHodgson** [comment]
>Yep, if the override models are set, the actual models will never load. You'll need to remove the overrides. The rotation of the controller models sounds like a known bug: the rotation will always be from the reference of the world's forward direction, based upon when you centered your space during mixed reality setup.

So even if I re-position my body and turn they'll never be in the correct position?  How do you launch without being too far from your initial frame of reference?

**Comment by keveleigh** [comment]
Yeah, I centered my device basically at my chair, so I always make sure to launch when facing in the same direction from a similar spot. A bit of a pain of a workaround for now, unfortunately. 

> So even if I re-position my body and turn they'll never be in the correct position?

Correct. For example, if you launch the app while facing 90 degrees to your right from the world's forward direction, your controllers will always be facing 90 degrees to the left.

**Comment by ElliotMebane** [comment]
This sounds like how Google's Daydream controller works. You have to *calibrate* it during your experience by pointing it towards the app's forward direction and pressing the calibration button for a couple of seconds. If the alignment drifts over time you re-calibrate. It's basically associating what the controller detects as North with what direction you've aligned your app to face.

**Comment by NeerajW** [comment]
@keveleigh can we please enumerate what current bugs you are investigating that are blocking merging of this PR? Let's get crisp on the list please.

Known bugs + bugs you are investigating fixes for.

**Comment by keveleigh** [comment]
No bugs (in our code) remain. I've addressed @StephenHodgson's and @NeerajW's feedback, and this PR should be good to go!

**Comment by StephenHodgson** [comment]
I'd like to test it again, but I never did get the controllers to show up correctly.

The test machine I had is gone now though, and I don't have the insider preview on my current machine.

**Comment by keveleigh** [comment]
@StephenHodgson Another reason you may not have seen them is that the Unity APIs currently return the controller position in the wrong frame of reference, so they render in/near the floor. This will be fixed in a future beta release though.

**Comment by StephenHodgson** [comment]
I didn't have a floor in my scene.  It was just an empty scene with the controllers.

**Comment by keveleigh** [comment]
Ahh interesting...

**Pull request merged**